File: project-euler/423/euler_423_v1.py
#!/usr/bin/env python3

# The Expat License
#
# Copyright (c) 2017, Shlomi Fish
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys

if sys.version_info > (3,):
    long = int
    xrange = range

MOD = 1000000007

s_n = 0

# a0 a1 a2 a3 | a4
# b0 b1 b2 b3 | b4
# S[b4] = S[a4] * 6 - (S[a4]-S[a3]) = 5*S[a4]+S[a3]
primes_s = open("primes.txt").read()
primes = [int(p) for p in (str(primes_s)).split("\n") if len(p) > 0] + [-1]


def base6(n):
    if n == 0:
        return ''
    return base6(n / 6) + str(n % 6)


sums = [6, 6, 6]
pi = 0
# PS = 1000
PS = 1
pa = PS
# for n in xrange(1,50+1):
for n in xrange(1, 50000000+1):
    # n_sums → new_sums
    n_sums = [(5*sums[0]) % MOD]
    for i in xrange(1, n+1):
        n_sums.append((sums[i]*5+sums[i-1]) % MOD)
    n_sums.append(n_sums[-1])
    print(["0[6]" + base6(x) for x in n_sums])
    # print ("n = ", n, " ; sums = " , [x//6 for x in sums])
    if primes[0] == n:
        primes.pop(0)
        pi += 1

    C = sums[pi]
    s_n += C
    s_n %= MOD
    if n == pa:
        pa += PS
        print("C(%d) = %d; S = %d" % (n, C, s_n))
    sums = n_sums

File: project-euler/423/euler_423_v1.cpp
// The Expat License
//
// Copyright (c) 2017, Shlomi Fish
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>

#define min(a,b) (((a)<(b))?(a):(b))

typedef long long LL;
typedef unsigned int ui;
#define MOD_PROTO 1000000007
const ui MOD = MOD_PROTO;

ui * lookup_5_prod_mods = NULL;
#define NUM_PRIMES_MIN_1 3001134
#define NUM_PRIMES (NUM_PRIMES_MIN_1 + 1)

int primes[NUM_PRIMES];
int * p_p;
int next_p;

ui s_n = 0;

#define lim 50000000
#define after_lim (lim+1)

ui one[NUM_PRIMES+1], two[NUM_PRIMES+1];

ui * current , * next, * temp;

const int DUMP_STEP = 10000;

int dump_at;

int main()
{
    {
        lookup_5_prod_mods = (ui *)malloc(sizeof(lookup_5_prod_mods[0]) * MOD);
        ui m = 0;
        for (ui i = 0 ; i < MOD; i++)
        {
            lookup_5_prod_mods[i] = m;
            if ((m += 5) >= MOD)
            {
                m -= MOD;
            }
        }
    }
    dump_at = DUMP_STEP;
    FILE * f = fopen ("primes.txt", "rt");

    p_p = primes;
    while (fscanf(f, "%d\n", p_p) == 1)
    {
        p_p++;
    }
    *(p_p) = -1;
    next_p = *(p_p = primes);

    fclose(f);

    memset(one, '\0', sizeof(one));
    memset(two, '\0', sizeof(two));

    one[0] = one[1] = one[2] = 6;
    current = one;
    next = two;
    int pi = 0;

    int n = 1;
    int prev_n = 1;
    while (1)
    {
        char fn[300];
        sprintf(fn, "DUMPS/%d.bin", dump_at);

        FILE * fh = fopen(fn, "rb");

        if (! fh)
        {
            n = prev_n;
            break;
        }
        fread (&s_n, sizeof(s_n), 1, fh);
        fread (current, sizeof(*current), sizeof(one)/sizeof(one[0]), fh);

        fclose(fh);

        dump_at += DUMP_STEP;
        prev_n = n;
        n = dump_at;
    }

    /* Restore p_p */
    for (int i = 1; i <= n; i++)
    {
        if (*(p_p) == i)
        {
            pi++;
            p_p++;
        }
    }
    next_p = *(p_p);
    for (; n < after_lim ; n++)
    {
        if (dump_at == n)
        {
            char fn[300];
            sprintf(fn, "DUMPS/%d.bin", dump_at);
            FILE * fh = fopen(fn, "wb");
            fwrite (&s_n, sizeof(s_n), 1, fh);
            fwrite (current, sizeof(*current), sizeof(one)/sizeof(one[0]), fh);
            fclose(fh);
            dump_at += DUMP_STEP;
        }
        next[0] = lookup_5_prod_mods[current[0]];
        const int my_top_idx = min(n, NUM_PRIMES_MIN_1);
        int i;
        for (i = 1; i <= my_top_idx; i++)
        {
            const ui my_next = lookup_5_prod_mods[current[i]] + current[i-1];
            next[i] = my_next >= MOD ? my_next - MOD : my_next;
        }
        next[i] = next[i-1];

        if (next_p == n)
        {
            pi++;
            next_p = *(++p_p);
        }

        const ui C = current[pi];
        if ((s_n += C) >= MOD)
        {
            s_n -= MOD;
        }
        std::cout << "C(" << n << ") = " << C << "; S = " << s_n << std::endl;

        {
            temp = current;
            current = next;
            next = temp;
        }
    }

    return 0;
}
